src/util/db/syncing.ts

```diff
@@ -0,0 +1,188 @@
+import { asMaybe } from 'cleaners'
+
+import { AppState } from '../../server'
+import { Checkpoint, CheckpointArray } from '../../types/checkpoints'
+import { asStoreFileDocument, StoreFileDocument } from '../../types/db-types'
+import { ChangeSetV2 } from '../../types/v2-types'
+import { Effect } from '../effect'
+import { trial } from '../trial'
+import { withRetries } from '../utils'
+import { equalCheckpoints, getCheckpointAt } from './checkpoints'
+import { checkDbResponseForErrors } from './error-checking'
+
+const MAX_DOCUMENT_VERSION_COUNT = 100
+
+export const writeUpdates = (appState: AppState) => (
+  repoId: string,
+  changeSet: ChangeSetV2
+): Promise<void> =>
+  withRetries(
+    async (): Promise<void> => {
+      const latestCheckpoint = await getCheckpointAt(appState)(repoId)
+
+      // Query for documents
+      const keys = Object.keys(changeSet).map(path => `${repoId}:${path}`)
+      const fetchResponse = await appState.dataStore.fetch({ keys })
+
+      // Prepare updated documents
+      const updatedStoreFileDocs = fetchResponse.rows.map((row, i) => {
+        const path = row.key.split(':')[1]
+        const box = changeSet[path]
+        const newVersion = latestCheckpoint.version + i + 1
+
+        if (row.error != null && row.error !== 'not_found') {
+          throw new Error(row.error)
+        }
+
+        const existingStoreFileDoc =
+          'doc' in row && row.doc != null
+            ? asMaybe(asStoreFileDocument)(row.doc)
+            : undefined
+        const templateStoreFileDoc = existingStoreFileDoc ?? {
+          _id: row.key,
+          box,
+          timestamp: Date.now(),
+          versions: []
+        }
+
+        return {
+          ...templateStoreFileDoc,
+          box,
+          timestamp: Date.now(),
+          versions: [newVersion, ...templateStoreFileDoc.versions].slice(
+            0,
+            MAX_DOCUMENT_VERSION_COUNT
```

Comment:
I just realized, if the version sum is the sum of each file document's `versions[0]`, then how would trimming affect the sum? Unless I misunderstand what you mean, I think we'd be safe.

Reply:
Client syncs to version 5, then goes away for a while. Some document, D, then has an array with [5,6,7,8,9,10] trimmed down to not include 5. Now, the client syncs again, and since version 5 is no longer on D, we don't include it in our fixups. Therefore, we think the client is -5 compared to where it should be.

If, at the same time, we have inconsistent files that happen to be worth +5, the client will sync successfully, and we will never realize the problem, leaving the client with a *permanent* inconsistent state.

The checkpoint syncing algorithm works reliably (QED) if the sums never go down, but deleting a historical version can make the sums go down, which means we need to do something special to avoid issues related to this.